pkg/object/websocketserver/proxy.go

```diff
@@ -0,0 +1,294 @@
+			if !handle() {
+				return
+			}
+		}
+	}
+}
+
+// run runs the websocket proxy.
+func (p *Proxy) run() {
+	spec := p.superSpec.ObjectSpec().(*Spec)
+	backendURL, err := url.Parse(spec.Backend)
+	if err != nil {
+		logger.Errorf("invalid websocketserver backend URL: %s ", spec.Backend)
+		return
+	}
+
+	p.backendURL = backendURL
+	dialer := defaultDialer
+	if strings.HasPrefix(spec.Backend, "wss") {
+		tlsConfig, err := spec.wssTLSConfig()
+		if err != nil {
+			logger.Errorf("gen websocketserver backend tls failed: %v, spec :%#v", err, spec)
+			return
+		}
+		dialer.TLSClientConfig = tlsConfig
+	}
+	p.dialer = dialer
+	p.upgrader = defaultUpgrader
+
+	http.HandleFunc("/", p.handle)
+	addr := fmt.Sprintf(":%d", spec.Port)
+	svr := &http.Server{
+		Addr:    addr,
+		Handler: nil,
+	}
+
+	if spec.HTTPS {
+		tlsConfig, err := spec.tlsConfig()
+		if err != nil {
+			logger.Errorf("gen websocketserver's httpserver tlsConfig: %#v, failed: %v", spec, err)
+		}
+		svr.TLSConfig = tlsConfig
+	}
+
+	if err := svr.ListenAndServe(); err != nil {
+		logger.Errorf("websocketserver ListenAndServe failed! err: %s\n", err.Error())
+	}
+}
+
+// copyHeader copies headers from the incoming request to the dialer and forward them to
+// the destinations.
+func (p *Proxy) copyHeader(req *http.Request) http.Header {
+
+	requestHeader := http.Header{}
+	if origin := req.Header.Get("Origin"); origin != "" {
+		requestHeader.Add("Origin", origin)
+	}
+	for _, prot := range req.Header[http.CanonicalHeaderKey("Sec-WebSocket-Protocol")] {
+		requestHeader.Add("Sec-WebSocket-Protocol", prot)
+	}
+	for _, cookie := range req.Header[http.CanonicalHeaderKey("Cookie")] {
+		requestHeader.Add("Cookie", cookie)
+	}
+	if req.Host != "" {
+		requestHeader.Set("Host", req.Host)
+	}
+
+	if clientIP, _, err := net.SplitHostPort(req.RemoteAddr); err == nil {
+		if prior, ok := req.Header["X-Forwarded-For"]; ok {
+			clientIP = strings.Join(prior, ", ") + ", " + clientIP
+		}
+		requestHeader.Set("X-Forwarded-For", clientIP)
+	}
+
+	requestHeader.Set("X-Forwarded-Proto", "http")
+	if req.TLS != nil {
+		requestHeader.Set("X-Forwarded-Proto", "https")
+	}
+
+	return requestHeader
+}
+
+// upgradeRspHeader passes only selected headers as return.
+func (p *Proxy) upgradeRspHeader(resp *http.Response) http.Header {
+	upgradeHeader := http.Header{}
+	if hdr := resp.Header.Get("Sec-Websocket-Protocol"); hdr != "" {
+		upgradeHeader.Set("Sec-Websocket-Protocol", hdr)
+	}
+	if hdr := resp.Header.Get("Set-Cookie"); hdr != "" {
+		upgradeHeader.Set("Set-Cookie", hdr)
+	}
+	return upgradeHeader
+}
+
+// handle implements the http.Handler that proxies WebSocket connections.
+func (p *Proxy) handle(rw http.ResponseWriter, req *http.Request) {
+	connBackend, resp, err := p.dialer.Dial(p.buildReq(req).String(), p.copyHeader(req))
+	if err != nil {
+		logger.Errorf("Proxy: couldn't dial to ws remote backend url: %s, err: %v", p.backendURL.String(), err)
```

Comment:
`logger.Errorf("%s dial %s failed: %v", p.superSpec().Name(), p.backendURL.String(), err)`. Please keep error message accurate and clean without redundant words.

Reply:
Already update to correct log format. Thx